File: src/main/java/com/enterprise/iqk/evaluation/EvaluationReportRenderer.java
package com.enterprise.iqk.evaluation;

import com.enterprise.iqk.evaluation.vo.EvalMetricSummaryVO;
import com.enterprise.iqk.evaluation.vo.EvalResultVO;
import com.enterprise.iqk.evaluation.vo.EvalRunVO;
import org.springframework.stereotype.Component;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

@Component
public class EvaluationReportRenderer {

    public String render(EvalRunVO run) {
        List<String> lines = new ArrayList<>();
        lines.add("# RAG Evaluation Report");
        lines.add("");
        lines.add("- Run ID: `" + run.getRunId() + "`");
        lines.add("- Dataset ID: `" + run.getDatasetId() + "`");
        lines.add("- Tenant: `" + run.getTenantId() + "`");
        lines.add("- Model Profile: `" + run.getModelProfile() + "`");
        lines.add("- Status: `" + run.getStatus() + "`");
        lines.add("- Generated At: " + LocalDateTime.now().format(DateTimeFormatter.ISO_LOCAL_DATE_TIME));
        lines.add("");
        lines.add("## Metrics");
        lines.add("");
        lines.add("| Metric | Value |");
        lines.add("| --- | ---: |");
        EvalMetricSummaryVO m = run.getMetrics();
        lines.add("| Run Score | " + pct(m.getRunScore()) + " |");
        lines.add("| Retrieval Hit Rate | " + pct(m.getRetrievalHitRate()) + " |");
        lines.add("| Citation Coverage | " + pct(m.getCitationCoverageRate()) + " |");
        lines.add("| Answer Faithfulness | " + pct(m.getAnswerFaithfulnessScore()) + " |");
        lines.add("| Avg Latency | " + String.format(Locale.ROOT, "%.1f ms", m.getAvgLatencyMs()) + " |");
        lines.add("| Failure Rate | " + pct(m.getFailureRate()) + " |");
        lines.add("");
        lines.add("## Cases");
        lines.add("");
        lines.add("| Case | Status | Score | Retrieval | Citation | Faithfulness | Latency |");
        lines.add("| --- | --- | ---: | ---: | ---: | ---: | ---: |");
        for (EvalResultVO result : run.getResults()) {
            lines.add("| `" + result.getCaseId() + "` | " + result.getStatus()
                    + " | " + pct(result.getScore())
                    + " | " + pct(result.getRetrievalHit())
                    + " | " + pct(result.getCitationCoverage())
                    + " | " + pct(result.getAnswerFaithfulness())
                    + " | " + result.getLatencyMs() + " ms |");
        }
        lines.add("");
        return String.join("\n", lines);
    }

    private String pct(double value) {
        return String.format(Locale.ROOT, "%.2f%%", value * 100.0);
    }
}

File: typescript/apps/api/src/evaluation/evaluation-report.renderer.ts
import { Injectable } from "@nestjs/common";

import type { EvaluationRunView } from "./evaluation.service.js";

@Injectable()
export class EvaluationReportRenderer {
  render(run: EvaluationRunView): string {
    const lines = [
      "# RAG Evaluation Report",
      "",
      `- Run ID: \`${run.runId}\``,
      `- Dataset ID: \`${run.datasetId}\``,
      `- Tenant: \`${run.tenantId}\``,
      `- Model Profile: \`${run.modelProfile}\``,
      `- Status: \`${run.status}\``,
      `- Generated At: ${localDateTime()}`,
      "",
      "## Metrics",
      "",
      "| Metric | Value |",
      "| --- | ---: |",
      `| Run Score | ${pct(run.metrics.runScore)} |`,
      `| Retrieval Hit Rate | ${pct(run.metrics.retrievalHitRate)} |`,
      `| Citation Coverage | ${pct(run.metrics.citationCoverageRate)} |`,
      `| Answer Faithfulness | ${pct(run.metrics.answerFaithfulnessScore)} |`,
      `| Avg Latency | ${run.metrics.avgLatencyMs.toFixed(1)} ms |`,
      `| Failure Rate | ${pct(run.metrics.failureRate)} |`,
      "",
      "## Cases",
      "",
      "| Case | Status | Score | Retrieval | Citation | Faithfulness | Latency |",
      "| --- | --- | ---: | ---: | ---: | ---: | ---: |",
      ...run.results.map((result) => [
        `| \`${result.caseId}\` | ${result.status}`,
        pct(result.score),
        pct(result.retrievalHit),
        pct(result.citationCoverage),
        pct(result.answerFaithfulness),
        `${result.latencyMs} ms |`
      ].join(" | ")),
      ""
    ];
    return lines.join("\n");
  }
}

function pct(value: number): string {
  return `${(value * 100).toFixed(2)}%`;
}

function localDateTime(date = new Date()): string {
  const offsetMs = date.getTimezoneOffset() * 60_000;
  return new Date(date.getTime() - offsetMs).toISOString().replace(/Z$/, "");
}
